libp2p/protocols/pubsub/gossipsub/behavior.nim

```diff
@@ -346,39 +355,46 @@ proc rebalanceMesh*(g: GossipSub, topic: string, metrics: ptr MeshMetrics = nil)
           grafts &= peer
 
   else:
-      trace "replenishing mesh outbound quota", peers = g.mesh.peers(topic)
-
-      var candidates = toSeq(
-        g.gossipsub.getOrDefault(topic, initHashSet[PubSubPeer]()) -
-        g.mesh.getOrDefault(topic, initHashSet[PubSubPeer]())
-      ).filterIt(
-        it.connected and
-        # get only outbound ones
-        it.outbound and
-        # avoid negative score peers
-        it.score >= 0.0 and
-        # don't pick explicit peers
-        it.peerId notin g.parameters.directPeers and
-        # and avoid peers we are backing off
-        it.peerId notin g.backingOff.getOrDefault(topic)
-      )
+    trace "replenishing mesh outbound quota", peers = g.mesh.peers(topic)
+
+    # this is a workaround to get the hashset from the table
+    # without creating a copy.
+    proc getCandidates(currentMesh: var HashSet[PubSubPeer]): seq[PubSubPeer] =
+      g.gossipSub.withValue(topic, peerList):
+        for it in peerList[]:
+          if
+              it.connected and
+              # get only outbound ones
+              it.outbound and
+              it notin currentMesh and
+              # avoid negative score peers
+              it.score >= 0.0 and
+              # don't pick explicit peers
+              it.peerId notin g.parameters.directPeers and
+              # and avoid peers we are backing off
+              it.peerId notin backingOff:
+            result.add(it)
+
+    var candidates =
```

Comment:
Same comment as in https://github.com/status-im/nim-libp2p/pull/708/files#r875077380